### DISCUSSION TITLE
firewall-and-vpn-without-root-on-android/

### DISCUSSION THREAD

**Discussion by giscus** [ItsIgnacioPortal/ItsIgnacioPortal.github.io#2] [Announcements]
# firewall-and-vpn-without-root-on-android/

No root or external servers needed. Requirements: Android 5.0+, 
446.03 MB of available storage, a computer and a cable to connect your phone.

https://itsignacioportal.github.io/firewall-and-vpn-without-root-on-android/

**Comment by flagel** [comment]
Nice guide - I'm glad someone is keeping this alive. :-)

Not important, just curious, is SOCKS5 with authentication working for you with the (as of now) latest Rethink? I believe it to be v055n.

**Reply by ItsIgnacioPortal** [reply]
Yes, SOCKS5 authentication does work on:
- RethinkDNS v0.5.5n-1-gc74a7ae1
- sing-box v1.9.3

RethinkDNS does fail to connect if you specify the wrong password when setting up the SOCKS5 proxy. This can be seen in the sing-box logs

**Reply by flagel** [reply]
Interesting. I am sure I set the password to something very simple when debugging to avoid PEBKAC. Thanks!

I was running Every Proxy, though, so maybe the error is on the side of Every Proxy. Then again, Every Proxy+SOCKS5 worked just fine with Rethink v054c.

**Reply by flagel** [reply]
Info for anyone trying to get Rethink to work in SOCKS5+auth-mode. That setup does not seem to work for everyone (me included), regardless if proxy used is sing-box or Every Proxy. More info: https://github.com/celzero/rethink-app/issues/919#issuecomment-2663880594

**Reply by ItsIgnacioPortal** [reply]
Hey @FLAGEL a new version of the guide is out now! I highly recommend you check it out: https://itsignacioportal.github.io/firewall-and-vpn-without-root-on-android-2026/

**Reply by flagel** [reply]
@ItsIgnacioPortal, nice work on the new guide! Hope to have a closer look at it in the future, and maybe let wireshark have a look at it too. ;-) Keep it up!

**Comment by slopsommelier** [comment]
Can singbox somehow bypass the problem with rethink where it can't properly route incoming LAN connections (e.g. localsend, kdeconnect)

**Reply by ItsIgnacioPortal** [reply]
No, I don't think so. Is there a github issue opened for that?

**Reply by slopsommelier** [reply]
I opened one. 

I was able to fix it by using invizible pro instead of rethink. 

You still need singbox, shelter/insular, and a wireguard VPN for the work profile part. 

You can allow/deny LAN access per app in invizible.

But invizible by itself can't do IP/port filtering, only app filtering. And the dev is kinda hostile toward any suggestions (I asked him to let me set the bootstrap resolvers in the dnscrypt proxy module to null or localhost since they are unnecessary and he acted like I was an idiot that didn't understand how dnscrypt-proxy works)

Granted, rethink forces you to select bootstrap resolvers but that's because it doesn't support stamps for DoH (yet, it's an open issue).

**Reply by revinco** [reply]
hi mvevitsis

I'm having the same issue and have no clue about Invizible setup. Rethink doesnt allow connection to my local network, although on another android smartphone I have the same problem connecting to my NAS via ProtonVPN. I allowed LAN access and activated split tunnelling, but no connection to my NAS possible.

If you find a solution or reask the open issue on git, I would appreciate. Otherwise a workaround added to the guide would be great.

**Reply by slopsommelier** [reply]
> hi mvevitsis
> 
> I'm having the same issue and have no clue about Invizible setup. Rethink doesnt allow connection to my local network, although on another android smartphone I have the same problem connecting to my NAS via ProtonVPN. I allowed LAN access and activated split tunnelling, but no connection to my NAS possible.
> 
> If you find a solution or reask the open issue on git, I would appreciate. Otherwise a workaround added to the guide would be great.

You just need to have invizible connected to singbox in work profile via a socks5 proxy. Then you can have the android killswitch enabled for both invizible and your VPN app inside the work profile and LAN apps on main profile will work. Enable invizible firewall and specify which apps you want to grant LAN access.

**Reply by revinco** [reply]
@ItsIgnacioPortal Could you please verify if this is the most practical solution and add to the blog how to set up with screenshots? Then we should have a full complete guide, if there is no other way to implement this VPN connection. Thank you.

**Reply by ItsIgnacioPortal** [reply]
> @ItsIgnacioPortal Could you please verify if this is the most practical solution and add to the blog how to set up with screenshots? Then we should have a full complete guide, if there is no other way to implement this VPN connection. Thank you.

I'll look into InviziblePro to see if it's a suitable replacement for RethinkDNS, then I'll run some tests, and if all goes well you should see an update on this blog by the end of next week 👍

**Reply by ItsIgnacioPortal** [reply]
Related issues:
- https://github.com/celzero/rethink-app/issues/1618
- https://github.com/celzero/rethink-app/issues/1725

**Reply by ItsIgnacioPortal** [reply]
This issue seems to have already been fixed in Rethink v054 (released on May 6th 2024). 

All that's needed is to enable "Do not route Private IPs" in Configure -> Network

See: https://github.com/celzero/rethink-app/issues/577#issuecomment-1674243033

**Reply by slopsommelier** [reply]
> This issue seems to have already been fixed in Rethink v054 (released on May 6th 2024). 
> 
> All that's needed is to enable "Do not route Private IPs" in Configure -> Network
> 
> See: https://github.com/celzero/rethink-app/issues/577#issuecomment-1674243033

That will not work if you have block connections without VPN enabled. You will simply block all private IPs.

**Reply by revinco** [reply]
> I opened one. 

Thanks, could you please post the git issue# ?

Has there been a commit resolve since you opened it?

**Reply by revinco** [reply]
@mvevitsis Could you please quote the pending git issue #?

**Reply by slopsommelier** [reply]
> @mvevitsis Could you please quote the pending git issue #?

https://github.com/celzero/rethink-app/issues/1618

It's not pending, it's closed and cannot technically be fixed unless they completely rewrite rethink. 

If this is your use case you must use a different app.

**Reply by ItsIgnacioPortal** [reply]
> > This issue seems to have already been fixed in Rethink v054 (released on May 6th 2024). 
> > 
> > All that's needed is to enable "Do not route Private IPs" in Configure -> Network
> > 
> > See: https://github.com/celzero/rethink-app/issues/577#issuecomment-1674243033
> 
> That will not work if you have block connections without VPN enabled. You will simply block all private IPs. 

That's to be expected. LAN traffic does not go through the VPN.

**Reply by slopsommelier** [reply]
> > > This issue seems to have already been fixed in Rethink v054 (released on May 6th 2024). 
> > > 
> > > All that's needed is to enable "Do not route Private IPs" in Configure -> Network
> > > 
> > > See: https://github.com/celzero/rethink-app/issues/577#issuecomment-1674243033
> > 
> > That will not work if you have block connections without VPN enabled. You will simply block all private IPs. 
> 
> That's to be expected. LAN traffic does not go through the VPN.

It can be done using invizible. 

For a diagram of the setup I describe, check the issue I linked above. I have spent weeks testing various configurations before I landed on this as it was the only way to have block connections enabled in both profiles AND use kdeconnect.

**Reply by revinco** [reply]
> > @ItsIgnacioPortal Could you please verify if this is the most practical solution and add to the blog how to set up with screenshots? Then we should have a full complete guide, if there is no other way to implement this VPN connection. Thank you.
> 
> I'll look into InviziblePro to see if it's a suitable replacement for RethinkDNS, then I'll run some tests, and if all goes well you should see an update on this blog by the end of next week 👍

Hi @ItsIgnacio, is Invizible then worth updating within your blog, for 'block connections without VPN' being enabled? Apparently it wouldn't block all private IPs and LAN, if we replace Singbox and RethinkDNS with it and it would be a perfect open source system (or as best possible)?

**Reply by ItsIgnacioPortal** [reply]
> Hi @ItsIgnacio, is Invizible then worth updating within your blog, for 'block connections without VPN' being enabled? Apparently it wouldn't block all private IPs and LAN, if we replace Singbox and RethinkDNS with it and it would be a perfect open source system (or as best possible)?

Yes, I won't be updating the guide for Invizible, since a solution is already available within Rethink.

**Reply by slopsommelier** [reply]
> > > @ItsIgnacioPortal Could you please verify if this is the most practical solution and add to the blog how to set up with screenshots? Then we should have a full complete guide, if there is no other way to implement this VPN connection. Thank you.
> > 
> > 
> > I'll look into InviziblePro to see if it's a suitable replacement for RethinkDNS, then I'll run some tests, and if all goes well you should see an update on this blog by the end of next week 👍
> 
> Hi @ItsIgnacio, is Invizible then worth updating within your blog, for 'block connections without VPN' being enabled? Apparently it wouldn't block all private IPs and LAN, if we replace Singbox and RethinkDNS with it and it would be a perfect open source system (or as best possible)?

You still need to use singbox

**Reply by slopsommelier** [reply]
> > Hi @ItsIgnacio, is Invizible then worth updating within your blog, for 'block connections without VPN' being enabled? Apparently it wouldn't block all private IPs and LAN, if we replace Singbox and RethinkDNS with it and it would be a perfect open source system (or as best possible)?
> 
> Yes, I won't be updating the guide for Invizible, since a solution is already available within Rethink.

There is no solution to use LAN apps while"block connections without VPN" enabled using Rethink. Invizible can do this because it passes private IPs through the VPN.

**Reply by ignoramous** [reply]
*rdns dev here*

> There is no solution to use LAN apps while"block connections without VPN" enabled using Rethink. Invizible can do this because it passes private IPs through the VPN.

Rethink passes "Private IPs" through it in ALL scenarios when *Block connections without VPN* (*VPN Lockdown* mode) is turned ON.

I don't know how InvziblePro makes LAN work in *VPN Lockdown* mode (the last time I had a chat with the developer Gdesh on this, they said Invizible doesn't do anything special and relies on the bundled NetGuard for all egress connections). 

As for LAN access over an actual VPN protocol like WireGuard ... that support is coming to Rethink by the end of this year, if all goes to plan.

**Reply by ItsIgnacioPortal** [reply]
@mvevitsis a new version of the guide is out now! I highly recommend you check it out: https://itsignacioportal.github.io/firewall-and-vpn-without-root-on-android-2026/

**Comment by abuturabofficial** [comment]
It went great, but after 5 mins internet stops working..... When I tried to reconnect got this:

Cloud warp: insular is interfering with UI

Proton VPN: more cryptic error that some other app is running as a always on

**Reply by abuturabofficial** [reply]
Turning Work profile off/on from QS panel shortcut of work profile fixed this issue for me.

**Reply by ItsIgnacioPortal** [reply]
> Turning Work profile off/on from QS panel shortcut of work profile fixed this issue for me.

Context: QS = Quick settings

**Reply by ItsIgnacioPortal** [reply]
@abuturabofficial a new version of the guide is out now! I highly recommend you check it out: https://itsignacioportal.github.io/firewall-and-vpn-without-root-on-android-2026/

**Comment by tsasam** [comment]
With this configuration, the DNS is not resolved by the VPN but by the systems' default (or whatever is chosen in RethinkDNS). Did someone manage to set up a DNS resolver in Sing-Box, so RethinkDNS can connect to it via SOCKS? (i.e. entering the IP+Port in DNS53

**Reply by tsasam** [reply]
I figured it out finally, with this code for Sing-Box and setting in RethinkDNS in DNS53 127.0.0.1:5353 as the server, the DNS is correctly going through the VPN:
```json
{
  "dns": {
    "servers": [
      {
        "type": "local",
        "tag": ""
      }      
    ]
  },
  "inbounds": [
    {
      "type": "socks",
      "tag": "",
      "listen": "127.0.0.1",
      "listen_port": 5353
    }
  ],
  "outbounds": [
    {
      "type": "direct",
      "tag": "direct"
    }
  ],
  "route": {
    "rules": [
      {
        "action": "sniff"
      },
      {
        "protocol": "dns",
        "action": "hijack-dns"
      },
      {
        "outbound": "direct"
      }
    ]
  }
}
```

**Reply by ItsIgnacioPortal** [reply]
The DNS queries always go through the VPN, and they're always the ones that you choose in RethinkDNS, never the system DNS. What are you trying to do here? Use the VPN's DNS servers instead of the ones you set in RethinkDNS?

**Reply by tsasam** [reply]
Yes, exactly. Previously it was going through the VPN as you said but did not use its DNS server. This way it can be used, depending on what is set in the RethinkDNS' DNS settings.

Using the VPN+DNS of the VPN  is important to reduce the fingerprint. Otherwise you will stand out using the VPNs IP but not its DNS resolver - like people directly using the VPN app would do.

Anyway, thanks for creating these tutorials, otherwise I would have never found out about this option! =)

**Comment by error-reporting** [comment]
hey the sing-box portion of this 2024 guide is the same as the 2025 guide except for the users fleid that has username and password why in the 2025 guide you tell users Do not try to adapt the 2025 guide to Rethink DNS?

**Reply by error-reporting** [reply]
just a question you don't have to change anything

**Reply by ItsIgnacioPortal** [reply]
So basically in the version that uses InviziblePRO, InviziblePRO does not support SOCKS5 proxy authentication when sending the DNS requests through the SOCKS5 proxy. So that's why we needed to remove that part from the sing-box configuration.

It was not much of a risk anyway because there is no reason that any app would want to connect to a local SOCKS5 proxy. There is also no risk of an external device connecting to that proxy because the proxy is only listening on the localhost interface.

As to why I tell people not to try to adapt the old guide to newer/different setups is because this setup has many variables and many applications that need to be configured independently in order to work with each other and there is just a lot of space for problems to occur. The risk with that is that if you configure something the wrong way, there is a real risk you might leak some of your private data and that is just not a risk I would advise anyone to take. But if you're willing to, you can try, however there is no guarantees if you try something that is outside of what I have recommended in the guides. You should test your custom setups thoroughly before relying on them to protect your data from bad actors.

**Reply by error-reporting** [reply]
I'm basically following your 2024 guide with the only difference is I removed the user field, the username and password section of sing-box but otherwise following fatefully to it
